add tox file for testing

sdague opened the pull request:
In order to make testing easier add a tox.ini file. This will let
folks run the bulk of what happens in travis pretty simply with just
typing:

  tox

Tox manages an independent virtualenv per target. Individual targets
can be run with 'tox -e py34' or 'tox -e lint'.

balloob commented:
We already have `script/test` that will run all needed tests (Oh, I see that it misses the requirements_all check, will add that later). I am not sure if I see the added value of adding Tox support.

On Travis the style and requirements_all validations are spread out over Python 3.4 and 3.5 to speed up the overall time.

sdague commented:
One of the advantages with tox is that it manages the venvs automatically for testing, which is nice when you are on Linux and have a lot of different python projects that require different things you are developing for.

balloob commented:
If people want to run in a virtuel environment (and we all should), they can just run `python3 -m venv .` and it's done. I think that if that is the added value of adding a `tox` file we shouldn't add it. It's another thing that has to be maintained etc.

sdague commented:
Ok, no prob. Abandoning.

balloob commented:
CC @kk7ds @flyte

kk7ds commented:
+1 for tox from me. Plus, @sdague is a wicked master at tox and I'm sure he'll help keep it up to date :)

flyte commented on tox.ini:
```diff
@@ -0,0 +1,18 @@
+commands = py.test {posargs}
+     flake8 --exclude www_static homeassistant
+; If you want to make tox run the tests with the same versions, create a
+; requirements.txt with the pinned versions and uncomment the following lines:
```
I don't understand what this comment means. The lines below are already uncommented?

MartinHjelmare commented:
Will this mean I could test the python 3.5 tests that happen on travis by just running tox? If so, I think this definitely adds value. I'm using python 3.4 in my venv for homeassitant as default. But then only running py.test won't check the python 3.5 things as I understand it.

balloob commented on tox.ini:
```diff
@@ -0,0 +1,18 @@
+setenv =
+    PYTHONPATH = {toxinidir}:{toxinidir}/homeassistant
+commands = py.test {posargs}
+     flake8 --exclude www_static homeassistant
```
`flake8` is a linter too.

balloob commented on tox.ini:
```diff
@@ -0,0 +1,18 @@
+[testenv:lint]
+basepython = python3
+commands =
+     pylint homeassistant
```
We should probably switch this to `script/lint` so we don't have to maintain the lint commands in two places.

balloob commented:
@MartinHjelmare the Python 3.5 tests can be run on Python 3.4 too. There are two that are Python 3.5 only: `script/lint` will run the linter and tell you about style violations. The other one is `script/gen_requirements_all.py validate`. Problem with that one is that it requires an environment that only includes the Home Assistant core dependencies, not the platform requirements.

MartinHjelmare commented:
@balloob Yes, so could we make the tox.ini to setup the proper environments for both 3.4 and 3.5 tests? I'd like to have an easy way of running all tests that travis does, but locally. I think that would help the devs and we should try to promote always running the tests locally before pushing. There's no mention about the test suite in https://github.com/balloob/home-assistant/blob/dev/CONTRIBUTING.md.

kk7ds commented:
@MartinHjelmare I don't think the reason you can't run everything local is a lack of py35, I think that the py35 job in travis is just *configured* to run more stuff than the 3.4 job:

https://github.com/balloob/home-assistant/blob/dev/script/cibuild#L8

MartinHjelmare commented:
@kk7ds Yes, well my point is that the local tests should mimic the travis tests by default, and if we provide an easy way of running the local tests, they will probably be used more.

sdague commented:
Well, learning what I have in the past month about tox and travis, I'd probably do this a bit different and have 3 tox targets. py34, py35, and lint. Travis can be configured to just run on tox aliases - https://github.com/sdague/arwn/blob/master/.travis.yml#L12-L15 

It would basically remove script/test, script/lint, as they would be in tox.ini. Travis and local would work exactly the same. Local would be about to do py34 & py35 & lint as long as you have both pythons available to you. Probably put the requirements check in the lint target. My understanding is the 35 / 34 split of non version specific tests was for time balancing.

I'm happy to work through this if people want to do it this way. It's the way I've done it on other projects. What do you think @balloob?

kk7ds commented:
@MartinHjelmare Yeah, agreed. A tox target of "cibuild" would be great to exactly replicate what runs in travis.

balloob commented:
That sounds great @sdague, I'm on board!

And yes, it was for time balancing

sdague committed:
convert testing infrastructure to tox

This converts the testing infrastructure to tox for both locally
testing and travis. This is nearly equivalent to the previous testing
with the only exception that linting fails with the first tool to fail
and won't process all of them.

Slightly tricky thing is that tox resets *all* of the environment for
it's subprocess runs by default. A couple of the dependencies we have
will not install in non UTF8 locales: temper-python & XBee.

sdague commented:
Ok this is the new attempt at things, after this merges testing will be:

> tox

Or if you want to call one of the specific language envs

> tox -e py34
> tox -e py35
> tox -e lint

balloob commented on tox.ini:
```diff
@@ -0,0 +1,26 @@
+commands =
+     flake8
+     pylint homeassistant
+     python script/gen_requirements_all.py validate
```
This command has to run without the dependencies being installed so it can catch when a person includes a dependency in the root of the module (because file import fails)

sdague replied:
I was pretty sure this is working as designed. We're trying to figure out requirements_all is out of sync, right? So having requirements_all.txt installed and missing something still makes it explode here (realistically the pylint also errors under those conditions).

If you have a scenario in mind this doesn't catch let me know.

sdague replied:
Also, on the sudo front, after dropping it we're still running on containers infrastructure. I think all projects that don't use a non container feature are on the containers infrastructure now.

sdague replied:
I can force linting to 3.5, that's easy enough.

balloob commented on tox.ini:
```diff
@@ -0,0 +1,26 @@
+     pylint homeassistant
+     python script/gen_requirements_all.py validate
+
+[flake8]
```
Wouldn't it be better if we put these in [`setup.cfg`](https://github.com/balloob/home-assistant/blob/dev/setup.cfg) ? That way linters in editors will be able to pick it up too

sdague replied:
Possibly, I've always seen it done in tox.ini and not setup.cfg, but I can move them over.

MartinHjelmare commented:
I think we should consider adding pep257 for flake8 linting to disallow new non pep257 docstrings. Anyone using having flake8-pep257 installed right now is getting scores of docstring errors when running the lint script. This lowers the confidence in the lint script and hides more serious issues. I think we should try to clean this up in the medium term.

sdague commented:
@MartinHjelmare I think that's a reasonable follow on. I'd also like to actually have flake8 run against our tests as well. But again, I think that's follow up.

kk7ds commented:
@sdague or use the "skip missing interpreters" flag right?

sdague commented:
@kk7ds right, good point. As long as we don't mess up the travis config that should be fine.

balloob commented:
I guess we will have to run the linting in Python 3.4 then after all?

balloob commented:
As back info, the reason why `script/gen_requirements_all.py` needs to not have all the requirements installed is because it is testing two things:
 - it validates that `requirements_all.txt` represents the dependencies mentioned in the components/platforms
 - to do the above it will load all components/platforms. By not having all requirements installed we are able to catch when a component or platform is importing a requirement in the module root (outside a method).

sdague commented:
@balloob for the lint and requirements targets tox, tox only specifies that it should be a python3 venv. Whatever your local platform defaults there will be used. In travis it'll run under 3.5 because the matrix is specified that way, it does look a little faster.

balloob commented:
Awesome. I think that it looks good :dolphin: . You can merge when you think it's done.

balloob commented on script/lint:
```diff
@@ -1,22 +0,0 @@
-# Run style checks
```
Can we keep this file and have it run the correct tox command?

sdague replied:
sure

balloob commented on script/test:
```diff
@@ -1,30 +0,0 @@
-#!/bin/sh
```
Can we keep this file and have it run the correct tox command?

sdague replied:
sure

sdague commented:
@balloob sounds good, I'll just work up a doc change to Contributing.md to explain testing with this change in place, and merge once I've got that.

balloob commented:
Can we for speed comparison do 1 run with `sudo: false` in `travis.yml`? It will allow us to use caching. https://docs.travis-ci.com/user/workers/container-based-infrastructure/

sdague commented:
@balloob sure, will set sudo: false in the travis.yaml, we can pull it out later if it's not really needed

kk7ds commented:
Looks good locally!

balloob commented:
Oh dang, the container infrastructure does not add any speed at all. Seems like the containers are slower.

Not in containers:
![screen shot 2016-02-14 at 10 09 21](https://cloud.githubusercontent.com/assets/1444314/13035356/16fdc1bc-d303-11e5-847e-e39a6d60e423.png)

Container infrastructure:
![screen shot 2016-02-14 at 10 09 28](https://cloud.githubusercontent.com/assets/1444314/13035355/16fd723e-d303-11e5-9e3f-eafae24793cf.png)

sdague commented:
Oh, interesting. I put travis up on my fork of HA, and because that's "new" in travis mind, it was on the containers infrastructure. But you are right, without sudo: false it doesn't look like the main repo goes to containers because it's older. That's fixed in this last round. So we're good to merge, but an interesting artifact.

kk7ds commented:
PR for developer index on next when this merges:

https://github.com/balloob/home-assistant.io/pull/216

balloob commented:
I realized that the screenshots that I posted were not fair - the container based ones had no cache yet. With cache warmed up the container infrastructure is faster for the actual tests (because installing dependencies is faster) but is still slower for linting (which is the slowest task).

![screen shot 2016-02-14 at 15 12 05](https://cloud.githubusercontent.com/assets/1444314/13037092/87a5da38-d32d-11e5-9054-39a804174eb2.png)
